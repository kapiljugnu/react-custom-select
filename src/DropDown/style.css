.multiselect {
    width: 200px;
  }
  
  .selectBox {
    position: relative;
  }
  
  .selectBox select {
    width: 100%;
    font-weight: bold;
  }
  
  .overSelect {
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
  }

  .caret {
    position: absolute;
    bottom: 8px;
    right: 5px;
    border: solid black;
    border-width: 0 1px 1px 0;
    padding: 2px;
    transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
  }
  
  .checkboxes {
    display: none;
    border: 1px #dadada solid;
    overflow: hidden;
    overflow-y: scroll;
    max-height: 100px;
  }
  
  .checkboxes label {
    display: block;
    text-align: left;
    margin-bottom: 5px;
    font-size:15px;
    user-select: none;
  }
  
  .checkboxes input[type="text"] {
    display: inherit;
    width: 100%;
    margin-bottom: 5px;
  }
  
  .checkboxes label:hover {
    background-color: #1e90ff;
  }
  