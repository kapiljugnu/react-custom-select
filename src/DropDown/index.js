import React from 'react';
import PropTypes from 'prop-types';

import './style.css';

class DropDown extends React.Component {
    state = {
        showOption: false,
        search: '',
        selectedOption: []
    }

    static propTypes = {
        options: PropTypes.array,
        selectTitle: PropTypes.string,
        searchTitle: PropTypes.string,
    }

    onDropDownClick = () => {
        this.setState(({ showOption }) => ({ showOption: !showOption }));
    }

    onOptionSelected = ({ target: { value, checked } }) => {
        const selected = this.state.selectedOption.filter(item => item !== value);
        if (checked) {
            selected.push(value);
        }
        this.setState({ selectedOption: selected })
    }

    onSearch = ({ target: { value } }) => {
        this.setState({ search: value });
    }

    render() {
        const { options, selectTitle, searchTitle } = this.props;
        const { search, showOption, selectedOption } = this.state;
        const checkboxStyle = { display: "none" };
        if (showOption) {
            checkboxStyle.display = "block";
        }
        return (
            <div className="multiselect">
                <div className="selectBox" onClick={this.onDropDownClick}>
                    <select>
                        <option>{selectTitle}</option>
                    </select>
                    <div className="overSelect" />
                </div>
                <div id="checkboxes" style={checkboxStyle}>
                    <input
                        type="text"
                        onChange={this.onSearch}
                        placeholder={searchTitle}
                    />
                    {options
                        .filter(item => item.toLowerCase().search(search.toLowerCase()) > -1)
                        .map(item => {
                            const props = {
                                type: "checkbox",
                                value: item,
                                onClick: this.onOptionSelected
                            };
                            if (selectedOption.includes(item)) {
                                props.checked = true;
                            }
                            return (
                                <label key={item}>
                                    <input {...props} />
                                    {item}
                                </label>
                            );
                        })}
                </div>
            </div>
        );
    }
}

export default DropDown;