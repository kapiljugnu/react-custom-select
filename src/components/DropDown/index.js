import React from 'react';
import PropTypes from 'prop-types';

import './style.css';
import SelectBox from './components/selectBox';
import CheckBoxes from './components/checkBoxes';

class DropDown extends React.Component {
    state = {
        showOption: false,
    }

    static propTypes = {
        options: PropTypes.array,
        selectTitle: PropTypes.string,
        searchTitle: PropTypes.string,
    }

    onDropDownClick = () => {
        this.setState(({ showOption }) => ({ showOption: !showOption }));
    }

    render() {
        const { options, selectTitle, searchTitle } = this.props;
        const { showOption } = this.state;
        const checkboxStyle = showOption ? { display: "block" } : { display: "none" };

        return (
            <div className="multiselect">
                <SelectBox selectTitle={selectTitle} onClick={this.onDropDownClick} />
                <CheckBoxes style={checkboxStyle} options={options} searchTitle={searchTitle} />
            </div>
        );
    }
}

export default DropDown;