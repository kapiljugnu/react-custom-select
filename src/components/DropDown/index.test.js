import React from 'react';
import DropDown from './index';
import renderer from 'react-test-renderer';

test('should match snapshot with hidden options', () => {
    const component = renderer.create(<DropDown
        options={["CandyEat", "OnionsPick", "FannyDuck", "Click-o-Wisp"]}
        selectTitle="Games"
        searchTitle="Search for games."
    />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('should match snapshot with visible options', () => {
    const component = renderer.create(<DropDown
        options={["CandyEat", "OnionsPick", "FannyDuck", "Click-o-Wisp"]}
        selectTitle="Games"
        searchTitle="Search for games."
    />);

    const selectBox = component.root.find(element => element.props.className === 'selectBox');
    selectBox.props.onClick();

    expect(component.toJSON()).toMatchSnapshot();
});
