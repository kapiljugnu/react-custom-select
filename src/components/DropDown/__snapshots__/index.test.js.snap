// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with hidden options 1`] = `
<div
  className="multiselect"
>
  <div
    className="selectBox"
    onClick={[Function]}
  >
    <select>
      <option>
        Games
      </option>
    </select>
    <div
      className="overSelect"
    >
      <div
        className="caret"
      />
    </div>
  </div>
  <div
    className="checkboxes"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <input
      onChange={[Function]}
      placeholder="Search for games."
      type="text"
    />
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="CandyEat"
      />
      CandyEat
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="OnionsPick"
      />
      OnionsPick
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="FannyDuck"
      />
      FannyDuck
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="Click-o-Wisp"
      />
      Click-o-Wisp
    </label>
  </div>
</div>
`;

exports[`should match snapshot with visible options 1`] = `
<div
  className="multiselect"
>
  <div
    className="selectBox"
    onClick={[Function]}
  >
    <select>
      <option>
        Games
      </option>
    </select>
    <div
      className="overSelect"
    >
      <div
        className="caret"
      />
    </div>
  </div>
  <div
    className="checkboxes"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <input
      onChange={[Function]}
      placeholder="Search for games."
      type="text"
    />
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="CandyEat"
      />
      CandyEat
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="OnionsPick"
      />
      OnionsPick
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="FannyDuck"
      />
      FannyDuck
    </label>
    <label>
      <input
        onClick={[Function]}
        type="checkbox"
        value="Click-o-Wisp"
      />
      Click-o-Wisp
    </label>
  </div>
</div>
`;
